core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,241 @@
+from trezor.crypto import hashlib
+from trezor.crypto.curve import ed25519
+from trezor.messages import CardanoAddressType
+
+from apps.common import cbor
+
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .helpers.utils import derive_public_key
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = dict[int, bytes]
+    CatalystRegistration = dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+CATALYST_VOTING_PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+
+METADATA_KEY_CATALYST_REGISTRATION = 61284
+METADATA_KEY_CATALYST_REGISTRATION_SIGNATURE = 61285
+
+
+def validate_auxiliary_data(
+    keychain: seed.Keychain,
+    auxiliary_data: CardanoTxAuxiliaryDataType | None,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.blob:
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.catalyst_registration_parameters:
+        _validate_catalyst_registration_parameters(
+            keychain,
+            auxiliary_data.catalyst_registration_parameters,
+            protocol_magic,
+            network_id,
+        )
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_auxiliary_data_blob(auxiliary_data_blob: bytes) -> None:
+    try:
+        # validation to prevent CBOR injection and invalid CBOR
+        # we don't validate data format, just that it's a valid CBOR
+        cbor.decode(auxiliary_data_blob)
+    except Exception:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_catalyst_registration_parameters(
+    keychain: seed.Keychain,
+    catalyst_registration_parameters: CardanoCatalystRegistrationParametersType | None,
```

Comment:
should not be optional, you're calling that only in the "if params is not falsey" branch